### PR TITLE
Add video annotation support

### PR DESCRIPTION
**Pull request by harshavardhana**
This work is still work in progress, doesn't seem to work for http://acroeng.adobe.com/Test_Files/classic_multimedia//VolvoS40V50-Full.pdf

It does seem like it worked in the beginning, but doesn't seem to work for me.  

Rebasing change from pull request #4376 to get some reviews and pitfalls.

### CODE REVIEW

**Comment by fbender** [comment]
@timvandermeij What's blocking merging this PR (except rebasing, ping @harshavardhana )?

**Comment by timvandermeij** [comment]
Nothing big I think. It needs a rebase and testing (unit tests especially).

**Comment by harshavardhana** [comment]
> Nothing big I think. It needs a rebase and testing (unit tests especially).

Can you tell me how would i go about adding those unit tests?

**Comment by harshavardhana** [comment]
> @timvandermeij What's blocking merging this PR (except rebasing, ping @harshavardhana )?

Rebased! and up for review.

File: src/core/annotation.js [hunk 1]
```diff
@@ -16,60 +16,12 @@
            stringToBytes, Promise, isArray, ObjectLoader, OperatorList,
            isValidUrl, OPS, createPromiseCapability, AnnotationType,
            stringToUTF8String, AnnotationBorderStyleType, ColorSpace,
-           AnnotationFlag, isInt */
+           AnnotationFlag, isInt, InteractiveAnnotation */
```
Comment (timvandermeij) [inline src/core/annotation.js]: This change can we removed: `InteractiveAnnotation` has been removed in #6178 as it was an unnecessary abstraction. Just inherit from `Annotation` directly.

File: src/core/annotation.js [hunk 2]
```diff
@@ -729,3 +681,63 @@ var LinkAnnotation = (function LinkAnnotationClosure() {
 
   return LinkAnnotation;
 })();
+
+var VideoAnnotation = (function VideoAnnotationClosure() {
+  function VideoAnnotation(params) {
+    InteractiveAnnotation.call(this, params);
```
Comment (timvandermeij) [inline src/core/annotation.js]: Same as above: replace `InteractiveAnnotation` with `Annotation`

File: src/core/annotation.js [hunk 3]
```diff
@@ -729,3 +681,63 @@ var LinkAnnotation = (function LinkAnnotationClosure() {
...
+      if (dict.get('Movie').has('Poster')) {
+        var poster = dict.get('Movie').get('Poster').getBytes();
+        var posterBlob = new Blob([poster]);
+        data.poster = PDFJS.createObjectURL(posterBlob);
+      }
+      data.src = dict.get('Movie').get('F').get('F');
+    }
+  }
+
+  Util.inherit(VideoAnnotation, InteractiveAnnotation, { });
```
Comment (timvandermeij) [inline src/core/annotation.js]: Same as above: replace `InteractiveAnnotation` with `Annotation`

**Comment by timvandermeij** [comment]
@harshavardhana I have added some review comments. You can add unit tests to https://github.com/mozilla/pdf.js/blob/master/test/unit/annotation_layer_spec.js. We create a new test suite per annotation type, so you can add the following:

<pre>
describe('VideoAnnotation', function() {
    ... // unit tests functions here
});
</pre>

At least we need some tests that the source of the video is interpreted correctly.

**Comment by harshavardhana** [comment]
> @harshavardhana I have added some review comments. You can add unit tests to https://github.com/mozilla/pdf.js/blob/master/test/unit/annotation_layer_spec.js. We create a new test suite per annotation type, so you can add the following:

Thanks for your suggestion @timvandermeij  - i will try to come up with something soon.

**Comment by humphd** [comment]
I was trying your branch, but what's on github doesn't seem to work with that example pdf--the worker dying on https://github.com/mozilla/pdf.js/blob/361fa83f3b6d502c80bdb0fc4cae71d857e80398/src/core/core.js#L276 with `AnnotationFactory` missing.  Do you have something locally that fixes this?

File: src/core/annotation.js [hunk 4]
```diff
@@ -21,55 +21,7 @@
...
-        return new WidgetAnnotation(parameters);
-
-      default:
-        warn('Unimplemented annotation type "' + subtype + '", ' +
-             'falling back to base annotation');
-        return new Annotation(parameters);
-    }
-  }
-};
+var SUPPORTED_TYPES = ['Link', 'Text', 'Widget', 'Screen', 'Movie'];
```
Comment (timvandermeij) [inline src/core/annotation.js]: This change is incorrect. The `AnnotationFactory` should not be removed (causing the effect that the user above describes), but rather it should be adapted to add a case for `Screen` and `Movie`.

File: src/display/annotation_helper.js [hunk 5]
```diff
@@ -12,8 +12,8 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-/* globals PDFJS, Util, AnnotationType, AnnotationBorderStyleType, warn,
-           CustomStyle, isExternalLinkTargetSet, LinkTargetStringMap */
+/* globals PDFJS, Util, AnnotationType, AnnotationBorderStyleType,
+           isExternalLinkTargetSet, LinkTargetStringMap, CustomStyle, warn */
```
Comment (timvandermeij) [inline src/display/annotation_helper.js]: Unnecessary change, please revert.

File: test/unit/annotation_layer_spec.js [hunk 6]
```diff
@@ -172,4 +172,8 @@ describe('Annotation layer', function() {
       expect(borderStyle.verticalCornerRadius).toEqual(0);
     });
   });
+  describe('VideoAnnotation', function() {
+    // unit tests functions here
```
Comment (timvandermeij) [inline test/unit/annotation_layer_spec.js]: This needs to be filled, but I think you were already aware of that. Just mentioning it for the sake of completeness.

**Comment by humphd** [comment]
I played with this some more, and with the `Annotation` ctor restored in this PR, I can see `VideoAnnotation`s being parsed correctly; however, the `getHtmlElementForVideoAnnotation` function in `annotation_helper.js` that should create the video element and usable UI, is never called.  I don't know the pdf.js internals enough to fix this, but I assume that the `VideoAnnotation` needs to somehow indicate that it has backing HTML and viewable content, such that the stuff in `pdfPage.getAnnotations()` should create and render the element.  Looking at how the `LinkAnnotation` (appears) to do it via `data.hasHtml = true;`, I wondered if this would be enough, but it isn't; I suspect something has to get parsed out of the annotation's data that isn't currently being done.  Any tips?

**Comment by timvandermeij** [comment]
`data.hasHtml = true;` should render the annotation and it is indeed absent in this PR, which should be changed. I have added some review comments above which might also contribute to the fact that the annotation is not rendered properly.

**Comment by humphd** [comment]
I've made all the changes discussed above (the `s/InteractiveAnnotation/Annotation/` stuff was already done), still no dice: https://github.com/humphd/pdf.js/commit/635c061ceb39d86077bfa46abc3d2f23ce7282c1.  @harshavardhana, I tried to send this to you as a PR, but github is being stupid and doesn't list you in the list of possible forks for pdf.js for some reason, so won't let me. @timvandermeij, what am I missing?

**Comment by harshavardhana** [comment]
> I've made all the changes discussed above (the s/InteractiveAnnotation/Annotation/ stuff was already done), still no dice: humphd@635c061. @harshavardhana, I tried to send this to you as a PR, but github is being stupid and doesn't list you in the list of possible forks for pdf.js for some reason, so won't let me. @timvandermeij, what am I missing?

I pulled in your changes and pushed here.

File: src/core/annotation.js [hunk 7]
```diff
@@ -14,8 +14,9 @@
  */
 /* globals PDFJS, Util, isDict, isName, stringToPDFString, warn, Dict, Stream,
            stringToBytes, Promise, isArray, ObjectLoader, OperatorList,
-           isValidUrl, OPS, AnnotationType, stringToUTF8String,
-           AnnotationBorderStyleType, ColorSpace, AnnotationFlag, isInt */
+           isValidUrl, OPS, createPromiseCapability, AnnotationType,
```
Comment (timvandermeij) [inline src/core/annotation.js]: The changes here should also be reverted.

File: src/core/annotation.js [hunk 8]
```diff
@@ -712,3 +717,65 @@ var LinkAnnotation = (function LinkAnnotationClosure() {
...
+      if (dict.get('Movie').has('Poster')) {
+        var poster = dict.get('Movie').get('Poster').getBytes();
+        var posterBlob = new Blob([poster]);
+        data.poster = PDFJS.createObjectURL(posterBlob);
+      }
+      data.src = dict.get('Movie').get('F').get('F');
+    }
+  }
+
+  Util.inherit(VideoAnnotation, Annotation, { });
```
Comment (timvandermeij) [inline src/core/annotation.js]: Nit: remove space between `{ }`

File: src/display/annotation_helper.js [hunk 9]
```diff
@@ -269,6 +269,45 @@ var AnnotationUtils = (function AnnotationUtilsClosure() {
     return container;
   }
 
+  function getHtmlElementForVideoAnnotation(item) {
+    var container = initContainer(item);
+    container.className = 'annotVideo';
+
+    container.style.borderColor = item.colorCssRgb;
```
Comment (timvandermeij) [inline src/display/annotation_helper.js]: The border styles are already handled, so I don't see why borders should be adjusted here.

File: src/shared/util.js [hunk 10]
```diff
@@ -45,7 +45,8 @@ var ImageKind = {
 var AnnotationType = {
   WIDGET: 1,
   TEXT: 2,
-  LINK: 3
+  LINK: 3,
```
Comment (timvandermeij) [inline src/shared/util.js]: Thanks to PR #6757 the changes in this file are no longer necessary as types `MOVIE` and `SCREEN` are now already available when you rebase this patch.

File: src/display/annotation_helper.js [hunk 11]
```diff
@@ -277,6 +316,8 @@ var AnnotationUtils = (function AnnotationUtilsClosure() {
         return getHtmlElementForTextAnnotation(data);
       case AnnotationType.LINK:
         return getHtmlElementForLinkAnnotation(data);
+      case AnnotationType.VIDEO:
```
Comment (timvandermeij) [inline src/display/annotation_helper.js]: After rebasing this patch, change this line to:

<pre>
case AnnotationType.MOVIE:
case AnnotationType.SCREEN:
  return ...
</pre>

File: src/display/annotation_helper.js [hunk 12]
```diff
@@ -269,6 +269,45 @@ var AnnotationUtils = (function AnnotationUtilsClosure() {
     return container;
   }
 
+  function getHtmlElementForVideoAnnotation(item) {
+    var container = initContainer(item);
```
Comment (timvandermeij) [inline src/display/annotation_helper.js]: After #6714 this function is now called `getContainer`.

File: src/core/annotation.js [hunk 13]
```diff
@@ -712,3 +717,65 @@ var LinkAnnotation = (function LinkAnnotationClosure() {
...
+      extName = extName.toLowerCase();
+      if (!extToMime.hasOwnProperty(extName)) {
+        return null;
+      }
+      return extToMime[extName];
+    }
+
+    var dict = params.dict;
+    var data = this.data;
+    data.annotationType = AnnotationType.VIDEO;
```
Comment (timvandermeij) [inline src/core/annotation.js]: This has to be either `AnnotationType.SCREEN` or `AnnotationType.MOVIE` depending on the type of annotation. I therefore think it is best to remove this class to `MovieAnnotation` and create a new class `ScreenAnnotation` that simply inherits everything from `MovieAnnotation` and only changes this `annotationType` variable. This makes more sense since both annotation types are different and Screen annotations might have other properties than Movie annotations, so having two classes is desirable.

File: src/display/annotation_helper.js [hunk 14]
```diff
@@ -269,6 +269,45 @@ var AnnotationUtils = (function AnnotationUtilsClosure() {
     return container;
   }
 
+  function getHtmlElementForVideoAnnotation(item) {
```
Comment (timvandermeij) [inline src/display/annotation_helper.js]: After #6770, all this code should be in a new class called `VideoAnnotationElement`. Please look at the current code to see how to structure this class. Creating an empty container, for instance, does not have to be done anymore as it is automatically available in `this.container` when you inherit from `AnnotationElement`.

**Comment by fbender** [comment]
@harshavardhana Are you still working on this? May you please address the review comments and rebase the branch?

**Comment by timvandermeij** [comment]
Closing as incomplete for the following reasons:

- This hasn't been worked on for a long time.
- The patch currently does not work for rendering video elements.
- The annotation code has been rewritten, so this needs to be fit into the new framework and therefore this patch is currently not in a mergeable state.

However, we will keep the issue open and link to this PR from there so that others may use this work as inspiration for a new PR.
